chore(docs): document training options and RuntimePatch API usage (#722)
kubeflow/sdk: docs/source/train/options.rst

Training Options
================

> Training options allow you to customize the metadata and configuration of your training jobs. They are optional arguments that can be passed to the ``options`` parameter of ``TrainerClient.train()``.

Reviewer: Existing `docs/source/train/*.rst` wrap at <=96 chars; this PR adds lines up to 308 (also lines 9, 34, 39, 47).

```suggestion
Training options allow you to customize the metadata and configuration of your training
jobs. They are optional arguments that can be passed to the ``options`` parameter of
``TrainerClient.train()``.
```

-----------------------------

The ``RuntimePatch`` option is a powerful mechanism for applying structured patches to the underlying ``TrainJob`` spec (specifically ``.spec.runtimePatches``) on Kubernetes. It enables you to customize pod templates, volumes, tolerations, node selectors, scheduling parameters, and container-level settings.

Structure of RuntimePatch
^^^^^^^^^^^^^^^^^^^^^^^^^

The ``RuntimePatch`` class contains a nested hierarchy of dataclasses representing different sections of the job specification:

* ``RuntimePatch``
   * ``training_runtime_spec`` (TrainingRuntimeSpecPatch)
      * ``template`` (JobSetTemplatePatch)
         * ``metadata`` (dict) - Metadata patches (labels, annotations) for the JobSet.
         * ``spec`` (JobSetSpecPatch)
            * ``replicated_jobs`` (list[ReplicatedJobPatch])
               * ``name`` (str) - Name of the replicated job to patch (e.g., ``"node"`` or ``"launcher"``).
               * ``template`` (JobTemplatePatch)
                  * ``metadata`` (dict)
                  * ``spec`` (JobSpecPatch)
                     * ``template`` (PodTemplatePatch)
                        * ``metadata`` (dict)

>                         * ``spec`` (PodSpecPatch)
>                            * ``node_selector`` (dict)
>                            * ``tolerations`` (list[dict])
>                            * ``volumes`` (list[dict])
>                            * ``containers`` (list[ContainerPatch])

Reviewer: `PodSpecPatch` has 10 fields (`options/kubernetes.py:109-118`); 6 are missing, which leaves the "scheduling parameters" claim above unsupported.

```suggestion
                        * ``spec`` (PodSpecPatch)
                           * ``service_account_name`` (str)
                           * ``volumes`` (list[dict])
                           * ``init_containers`` (list[ContainerPatch])
                           * ``containers`` (list[ContainerPatch])
                           * ``image_pull_secrets`` (list[dict])
                           * ``security_context`` (dict)
                           * ``node_selector`` (dict[str, str])
                           * ``affinity`` (dict)
                           * ``tolerations`` (list[dict])
                           * ``scheduling_gates`` (list[dict])
```

>                               * ``name`` (str) - Name of the container to patch (e.g., ``"pytorch"``).

Reviewer: The trainer container is named `node`, not `pytorch` (`constants.py:88`, looked up at `backends/kubernetes/utils.py:104`). `"pytorch"` is the Training Operator v1 `PyTorchJob` name.

```suggestion
                              * ``name`` (str) - Name of the container to patch. Must exist in the Runtime (e.g. ``"node"``).
```

>                               * ``env`` (list[dict])

Reviewer: Upstream forbids `env` here for the `node`/initializer containers (`TrainerV1alpha1ContainerPatch.env` description).

```suggestion
                              * ``env`` (list[dict]) - Not allowed for the ``node``, ``dataset-initializer``, or ``model-initializer`` containers; use ``CustomTrainer(env=...)`` or the Initializer API instead.
```

                              * ``volume_mounts`` (list[dict])
                              * ``security_context`` (dict)

Advanced Patching Example
^^^^^^^^^^^^^^^^^^^^^^^^^

> Here is a complete example of using ``RuntimePatch`` to mount a PVC to the training container, set custom environment variables, configure a node selector, and add a scheduling toleration:

Reviewer: Drop the env claim (not settable via `ContainerPatch` here) and wrap to ~100 chars.

```suggestion
Here is a complete example of using ``RuntimePatch`` to mount a PVC to the training
container, configure a node selector, add a scheduling toleration, and set a container
security context:
```

                                                   "key": "gpu-taint",
                                                   "operator": "Exists",
                                                   "effect": "NoSchedule",
                                               }
                                           ],
                                           # Define volumes
                                           volumes=[
                                               {
                                                   "name": "data-volume",
                                                   "persistentVolumeClaim": {
                                                       "claimName": "mnist-pvc"
                                                   }
                                               }
                                           ],
                                           # Patch the training container
                                           containers=[
                                               ContainerPatch(
                                                   name="pytorch",
                                                   # Mount the volume

>                                                    volume_mounts=[
>                                                        {
>                                                            "name": "data-volume",
>                                                            "mountPath": "/mnt/data",
>                                                        }
>                                                    ],
>                                                    # Set custom environment variables
>                                                    env=[
>                                                        {
>                                                            "name": "DATA_DIR",
>                                                            "value": "/mnt/data",
>                                                        }
>                                                    ],
>                                                    # Configure container security settings
>                                                    security_context={
>                                                        "privileged": True,
>                                                    },
>                                                )
>                                            ],

Reviewer: Three fixes: `pytorch` is not a container in the Runtime (it is `node`); `env` is rejected by the API for the `node` container; and docs should not showcase `privileged: True`.

```suggestion
                                           containers=[
                                               ContainerPatch(
                                                   # Must match a container in the
                                                   # Runtime; "node" is the trainer.
                                                   name="node",
                                                   # Mount the volume
                                                   volume_mounts=[
                                                       {
                                                           "name": "data-volume",
                                                           "mountPath": "/mnt/data",
                                                       }
                                                   ],
                                                   # Harden the container
                                                   security_context={
                                                       "allowPrivilegeEscalation": False,
                                                       "capabilities": {"drop": ["ALL"]},
                                                   },
                                               )
                                           ],
```

                                       )
                                   )
                               )
                           ),
                       )
                   ]
               )
           )
       )
   )

   # Submit training job with the patch
   client.train(
       trainer=CustomTrainer(func=train_fn),
       options=[patch]

>    )

Reviewer: Keeps `DATA_DIR` in the example via the supported path (`CustomTrainer.env`, `types.py:62`).

```suggestion
   # Submit training job with the patch. Env vars for the "node" container must be
   # set on the trainer, not via ContainerPatch.
   client.train(
       trainer=CustomTrainer(func=train_fn, env={"DATA_DIR": "/mnt/data"}),
       options=[patch],
   )
```

.. autoclass:: kubeflow.trainer.options.ContainerPatch
   :members:
   :show-inheritance:

> 

Reviewer: The file now ends with a blank line; the `end-of-file-fixer` pre-commit hook will fail CI.

```suggestion
   :show-inheritance:
```
